File: app/src/main/java/me/ykrank/s1next/view/fragment/BaseFragment.java
package me.ykrank.s1next.view.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.view.View;

import com.google.common.base.Optional;
import com.squareup.leakcanary.RefWatcher;

import java.lang.ref.WeakReference;

import me.ykrank.s1next.App;
import me.ykrank.s1next.AppComponent;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.UserValidator;
import me.ykrank.s1next.view.internal.CoordinatorLayoutAnchorDelegate;

public abstract class BaseFragment extends Fragment {

    protected UserValidator mUserValidator;

    protected CoordinatorLayoutAnchorDelegate mCoordinatorLayoutAnchorDelegate;
    @Nullable
    protected WeakReference<Snackbar> mRetrySnackbar;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        mCoordinatorLayoutAnchorDelegate = (CoordinatorLayoutAnchorDelegate) context;
    }

    @Override
    @CallSuper
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        AppComponent appComponent = App.getAppComponent(getContext());
        mUserValidator = appComponent.getUserValidator();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        RefWatcher refWatcher = App.get().getRefWatcher();
        refWatcher.watch(this);
    }

    @Override
    public void onDetach() {
        super.onDetach();

        mCoordinatorLayoutAnchorDelegate = null;
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);

        // see http://stackoverflow.com/a/9779971
        if (isVisible() && !isVisibleToUser) {
            // dismiss retry Snackbar when current Fragment hid
            // because this Snackbar is unrelated to other Fragments
            dismissRetrySnackbarIfExist();
        }
    }

    public void showRetrySnackbar(CharSequence text, View.OnClickListener onClickListener) {
        Optional<Snackbar> snackbar = mCoordinatorLayoutAnchorDelegate.showLongSnackbarIfVisible(
                text, R.string.snackbar_action_retry, onClickListener);
        if (snackbar.isPresent()) {
            mRetrySnackbar = new WeakReference<>(snackbar.get());
        }
    }
    
    protected void showShortSnackbar(CharSequence text){
        mCoordinatorLayoutAnchorDelegate.showShortSnackbar(text);
    }
    
    protected void showShortSnackbar(@StringRes int resId){
        mCoordinatorLayoutAnchorDelegate.showShortSnackbar(resId);
    }

    protected void showShortText(@StringRes int resId){
        mCoordinatorLayoutAnchorDelegate.showShortText(getString(resId));
    }

    protected void dismissRetrySnackbarIfExist() {
        if (mRetrySnackbar != null) {
            Snackbar snackbar = mRetrySnackbar.get();
            if (snackbar != null && snackbar.isShownOrQueued()) {
                snackbar.dismiss();
            }
            mRetrySnackbar = null;
        }
    }

}

File: app/src/main/java/me/ykrank/s1next/view/fragment/BaseFragment.kt
package me.ykrank.s1next.view.fragment

import android.app.Activity
import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.design.widget.Snackbar
import android.view.View
import android.widget.Toast
import com.github.ykrank.androidtools.ui.LibBaseFragment
import me.ykrank.s1next.App
import me.ykrank.s1next.data.User
import me.ykrank.s1next.data.api.UserValidator
import me.ykrank.s1next.util.ErrorUtil
import me.ykrank.s1next.view.activity.BaseActivity
import javax.inject.Inject

abstract class BaseFragment : LibBaseFragment() {

    @Inject
    internal lateinit var mUserValidator: UserValidator
    @Inject
    internal lateinit var mUser: User

    @CallSuper
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        App.appComponent.inject(this)
    }

    fun showRetrySnackbar(throwable: Throwable, onClickListener: View.OnClickListener) {
        val context = context ?: return
        showRetrySnackbar(ErrorUtil.parse(context, throwable), onClickListener)
    }

    protected fun showShortSnackbar(throwable: Throwable) {
        val context = context ?: return
        showShortSnackbar(ErrorUtil.parse(context, throwable))
    }

    /**
     * If current [android.app.Activity] is visible, sets result message to [Activity]
     * in order to show a short [Snackbar] for message during [.onActivityResult],
     * otherwise show a short [android.widget.Toast].

     * @param text The text to show.
     * *
     * @see BaseActivity.onActivityResult
     */
    protected fun showShortTextAndFinishCurrentActivity(text: CharSequence?) {
        val activity = activity ?: return
        val app = activity.applicationContext as App
        // Because Activity#onActivityResult(int, int, Intent) is always invoked when current app
        // is running in the foreground (so we are unable to set result message to Activity to
        // let BaseActivity to show a Toast if our app is running in the background).
        // So we need to handle it by ourselves.
        if (app.isAppVisible) {
            BaseActivity.setResultMessage(activity, text)
        } else {
            Toast.makeText(app, text, Toast.LENGTH_SHORT).show()
        }

        activity.finish()
    }
}
